### ISSUE TITLE
Question about speech to text quotas

### ISSUE THREAD

**Issue by lfcnassif** [googleapis/java-speech#171]
Hi, I have already read the docs about quotas for speech to text. They say 900 requests per minute and I see that in my google patform account. But I am getting 
```
RESOURCE_EXHAUSTED: Quota exceeded for quota metric 'Requests' and limit 'Requests per minute'
```
before processing 450 small audios and much before 1min, sometimes at the beggining. I tried to throttle to 15 requests per second, same result.

And while long audios are being processed, they still use a quota slot? Or just the instant when they are sent is used when computing requests per minute or per second?

**Comment by danoscarmike** [comment]
@lfcnassif thanks for reaching out. Transferring your issue to java-speech where it should get the right eyes on it.

**Comment by munkhuushmgl** [comment]
@lfcnassif Are your audio files stored in Google Cloud Storage or reading from your local machine ?

What are the types of your audio files?

What's maximum length of single audio of your 450 small audio files?

For per your request,

- Is your single request contains less than and equals to 100,000 chars?
- Is your single request contains less than and equals to 5000 phrases (each phrase contains at max 100 chars) ?

**Comment by lfcnassif** [comment]
Hi @munkhuushmgl, thank you for your attention.
- audios are all stored locally
- they all are the same file copied multiple times: wav pcm16le 16khz with 16bits per sample, 2s duration.
- when it works, recognized text is "gatinha"

As I said, often the exception is thrown much before 450 audios are processed, just with a few dozens...

**Comment by lfcnassif** [comment]
Any thoughts?

**Comment by munkhuushmgl** [comment]
I have talked with one of engineers for Speech but still have not the answer for it.

Here is what I got so far:

1. And while long audios are being processed, they still use a quota slot? Or just the instant when they are sent is used when computing requests per minute or per second? 

> **they should not use a quota "slot" while the audio is being processed**

2. 900 requests per minute

>  still waiting for response from speech team.

**Comment by b-loved-dreamer** [comment]
Hi, @lfcnassif 

- based on your specific use case, what would I need to do to recreate this issue? 

- When you say "they all are the same file copied multiple times", do you mean you have 450 copies of the same file OR you're sending (trying) 450 requests with the same file every time?

Thank you so much for being patient.

**Comment by lfcnassif** [comment]
Hi @b-loved-dreamer,

My steps to reproduce exception above back in April: pick up a small audio (around 2s or 3s), copy it 900 times in a local folder and send each copy to java speech-to-text service from parallel threads (24 threads in my configuration). Around 450 audios sent, sometimes before, sometimes after, RESOURCE_EXHAUSTED begins to be thrown.

**Comment by lfcnassif** [comment]
below is some code snipet extracted from a bigger class called by my parallel threads to send audios for transcription:
```
protected class TextAndScore {
        String text;
        double score;
 }

protected TextAndScore transcribeWavPart(File tmpFile) throws Exception {

        // google has request per time quota limits
        synchronized (lock) {
            long t = System.currentTimeMillis();
            long dif = t - lastTime;
            if (dif < requestIntervalMillis) {
                Thread.sleep(requestIntervalMillis - dif);
            }
            lastTime = System.currentTimeMillis();
        }

        TextAndScore textAndScore = null;
        try {
            // The language of the supplied audio
            String languageCode = languages.get(0);

            List<String> alternativeLangs = languages.subList(1, languages.size());

            Builder builder = RecognitionConfig.newBuilder().setLanguageCode(languageCode)
                    .addAllAlternativeLanguageCodes(alternativeLangs);

            RecognitionConfig config = builder.build();

            ByteString content;
            try (InputStream is = Files.newInputStream(tmpFile.toPath())) {
                content = ByteString.readFrom(is);
            }
            RecognitionAudio audio = RecognitionAudio.newBuilder().setContent(content).build();

            LongRunningRecognizeRequest request = LongRunningRecognizeRequest.newBuilder().setConfig(config)
                    .setAudio(audio).build();

            OperationFuture<LongRunningRecognizeResponse, LongRunningRecognizeMetadata> future = speechClient
                    .longRunningRecognizeAsync(request);

            LongRunningRecognizeResponse response = future
                    .get(MIN_TIMEOUT + (timeoutPerSec * tmpFile.length() / WAV_BYTES_PER_SEC), TimeUnit.SECONDS);

            StringBuilder text = new StringBuilder();
            float confidence = 0;
            int i = 0;
            for (SpeechRecognitionResult result : response.getResultsList()) {
                // First alternative is the most probable result
                SpeechRecognitionAlternative alternative = result.getAlternativesList().get(0);
                confidence += alternative.getConfidence();
                i++;
                text.append(alternative.getTranscript()).append(" ");
            }
            if (i == 0)
                i = 1;

            textAndScore = new TextAndScore();
            textAndScore.text = text.toString();
            textAndScore.score = confidence / i;

            LOGGER.debug("GG Transcript of {} : {}", evidence.getPath(), text.toString());

        } catch (Exception e) {
            LOGGER.error("Failed to transcript {} : {}", evidence.getPath(), e);
            LOGGER.warn("", e);
        }

        return textAndScore;

    }
```

lock is a static Object instance

**Comment by b-loved-dreamer** [comment]
Hi @lfcnassif 

I was not able to repro the issue you're describing. I am raising this issue with our product team. In the mean time, can you try to accomplish your goal using our non-beta libraries?

**Comment by b-loved-dreamer** [comment]
Won't fix (Can't repro)

**Comment by lfcnassif** [comment]
I'm returning to this issue, I'm really sorry for the long time elapsed... Just tried the non beta classes of google-cloud-speech-1.22.5 (same version used before), same problem happened. I also tried to change the end point according to https://cloud.google.com/speech-to-text/docs/endpoints with no success. I should state the `RESOURCE_EXHAUSTED: Quota exceeded for quota metric 'Requests' and limit 'Requests per minute'` is thrown after about 38 seconds sending requests and before 150 audios are processed, so I think I'm far away from the 900 audios per minute quota limit. I'm sending requests in parallel, should I wait a request response after sending another request? 

Since then, I have recommended using other speech-to-text provider/plugin for my application users because of this issue. But some of them are interested in Google's service, because it has higher quality with some audios, depending on accents. So I'm interested in solving this again and I'm willing to provide needed information and run tests to investigate this.

Thank you for your time,
Luis

**Comment by lfcnassif** [comment]
Just to let you know, I'm comparing some pt-BR models from different providers/authors on different data sets here:
https://github.com/sepinf-inc/IPED/issues/1214#issuecomment-1206666125

I don't know what changed in your service, but this quota problem is gone. I haven't change anything in my code, so probably it was some issue with Google service computing wrong quota metrics, maybe with the brazilian end point, as I suggested before...

**Comment by lfcnassif** [comment]
PS: the only difference is that I created another account, service and used different credentials, but the old one seemed good to me, since the pricing type and quota were the same.
